src/executables/util_cs.c

```diff
@@ -3994,6 +3994,827 @@ tde (UTIL_FUNCTION_ARG * arg)
+
+      process_column_data (&ptr, &sql, &max_sql_size, attr->type->id);
+      if (i != num_change_col - 1)
+	{
+	  strcat (sql, " and ");
+	}
+      else
+	{
+	  strcat (sql, " limit 1");
+	}
+
+      attr = attr->order_link;
+    }
+
+  if (is_detail)
+    {
+      ptr = *loginfo;
+
+      original_sql = (char *) malloc (max_original_size);
+      if (original_sql == NULL)
+	{
+	  util_log_write_errid (MSGCAT_UTIL_GENERIC_NO_MEM);
+	  error = ER_OUT_OF_VIRTUAL_MEMORY;
+	  goto error;
+	}
+
+      sprintf (original_sql, "insert into [%s] values ( ", classname);
+
+      ptr = or_unpack_int (ptr, &num_change_col);
+      for (i = 0; i < num_change_col; i++)
+	{
+	  ptr = or_unpack_int (ptr, &change_index);
+	}
+
+      attr = class_->ordered_attributes;
+      for (i = 0; i < num_change_col; i++)
+	{
+	  /* check SQL length
+	   * cond_sql + length of ", " */
+	  error = check_and_resize_sql_memory (&original_sql, strlen (original_sql) + 3, &max_original_size);
+	  if (error != NO_ERROR)
+	    {
+	      goto error;
+	    }
+
+	  process_column_data (&ptr, &original_sql, &max_original_size, attr->type->id);
+	  if (i != num_change_col - 1)
+	    {
+	      strcat (original_sql, ", ");
+	    }
+	  else
+	    {
+	      strcat (original_sql, " )");
+	    }
+
+	  attr = attr->order_link;
+	}
+    }
+
+  ptr = or_unpack_int (ptr, &num_cond_col);
+
+  *loginfo = ptr;
+
+  if (is_detail)
+    {
+      print_flashback_detail (trid, user, sql, original_sql, outfp);
+    }
+  else
+    {
+      fprintf (outfp, "%s\n", sql);
+    }
+
+  free_and_init (sql);
+  if (original_sql != NULL)
+    {
+      free_and_init (original_sql);
+    }
+
+  return NO_ERROR;
+
+error:
+  if (sql != NULL)
+    {
+      free_and_init (sql);
+    }
+
+  if (original_sql != NULL)
+    {
+      free_and_init (original_sql);
+    }
+
+  return error;
+}
+
+typedef enum
+{
+  FLASHBACK_INSERT = 0,
+  FLASHBACK_UPDATE,
+  FLASHBACK_DELETE
+} FLASHBACK_DML_TYPE;
+static int
+print_flashback_info (char *loginfo, int num_item, dynamic_array * classlist, OID * oidlist, bool is_detail,
+		      FILE * outfp)
+{
+  int length;
+  TRANID trid;
+  char *user;
+  int dataitem_type;
+
+  int error = NO_ERROR;
+
+  FLASHBACK_DML_TYPE dml_type;
+
+  INT64 b_classoid;
+  OID classoid;
+  int class_index = 0;
+  char classname[SM_MAX_IDENTIFIER_LENGTH] = "\0";
+
+  char *ptr = loginfo;
+
+  for (int i = 0; i < num_item; i++)
+    {
+      ptr = PTR_ALIGN (ptr, MAX_ALIGNMENT);
+      ptr = or_unpack_int (ptr, &length);
+      ptr = or_unpack_int (ptr, &trid);
+      ptr = or_unpack_string_nocopy (ptr, &user);
+      ptr = or_unpack_int (ptr, &dataitem_type);
+
+      ptr = or_unpack_int (ptr, (int *) &dml_type);
+      ptr = or_unpack_int64 (ptr, &b_classoid);
+      memcpy (&classoid, &b_classoid, sizeof (OID));
+
+      class_index = get_class_index (da_size (classlist), oidlist, classoid);
+
+      da_get (classlist, class_index, classname);
+
+      /* TODO: make enum for dml type */
```

Comment:
Seems that you already have made enum.

Reply:
Thanks, it was progress of applying feedback. I will remove it